### ISSUE TITLE
[Question] Has gated attention been tested on ViT architectures?

### ISSUE THREAD

**Issue by ILoveSorasakiHina** [qiuzh20/gated_attention#11]
Hi, thank you for your great work!
I'm curious whether your team has conducted any experiments applying Gated Attention to Vision Transformers (ViT)?
I implemented a Gated Multi-Head Attention module based on your paper and codebase:
```python
import torch
import torch.nn as nn
import torch.nn.functional as F
class GatedMultiHeadAttention(nn.Module):
    def __init__(self, embed_dim, num_heads):
        super().__init__()
        self.embed_dim = embed_dim
        self.num_heads = num_heads
        self.head_dim = embed_dim // num_heads
        
        self.qkv = nn.Linear(embed_dim, 3 * embed_dim)
        self.out_proj = nn.Linear(embed_dim, embed_dim)
        
        # Gating mechanism
        self.gate = nn.Linear(embed_dim, embed_dim)
        
    def forward(self, x):
        B, L, D = x.shape
        
        # 1. QKV projection
        qkv = self.qkv(x).reshape(B, L, 3, self.num_heads, self.head_dim)
        qkv = qkv.permute(2, 0, 3, 1, 4)
        q, k, v = qkv[0], qkv[1], qkv[2]
        
        # 2. Scaled dot-product attention
        attn_out = F.scaled_dot_product_attention(q, k, v)
        # Shape: [B, num_heads, L, head_dim]
        
        # 3. Compute gate scores
        gate_scores = torch.sigmoid(self.gate(x))
        gate_scores = gate_scores.reshape(B, L, self.num_heads, self.head_dim)
        gate_scores = gate_scores.permute(0, 2, 1, 3)
        # Shape: [B, num_heads, L, head_dim]
        
        # 4. Apply gating
        gated_out = attn_out * gate_scores
        
        # 5. Concatenate heads and project
        gated_out = gated_out.permute(0, 2, 1, 3).reshape(B, L, D)
        output = self.out_proj(gated_out)
        
        return output
```
I tested this implementation on my own dataset, but didn't observe significant improvements. I would appreciate it if you could help verify whether my implementation is correct, or if there are any known considerations when applying this approach to vision tasks.

Thanks in advance!

**Comment by qiuzh20** [comment]
Thank you for your question! We primarily evaluated the effectiveness of gating under the LLM pretraining setting and did not validate it in computer vision–related experiments.

Regarding the limited performance improvement, we have the following hypotheses:

1. When the model size is relatively small—particularly when the number of attention heads is limited and multi-head attention (MHA) is used instead of grouped-query attention (GQA)—the low-rank issue between \( W_v \) and \( W_o \) may not be severe. Consequently, the benefit from enhanced nonlinearity could be modest. In this case, we suggest trying v-norm or applying group normalization to the attention output and observing whether the training loss changes. If no noticeable change occurs, it likely indicates that enhancing nonlinearity offers limited gains under your experimental setup.

2. When training is insufficient, phenomena such as attention sinks and massive activations may not be prominent, which means another potential benefit of gating—mitigating these issues—might not manifest. You could examine your base model to see whether attention sinks or massive activations occur (similar to the observations reported in the "Vision Transformers Need Registers" paper).

**Comment by ILoveSorasakiHina** [comment]
Thank you for your detailed response! 

I will try the approaches you recommended.

If I discover any new findings, I will follow up with an update. Thank you again for your guidance!

**Comment by Zhangquan-se** [comment]
> Hi, thank you for your great work! I'm curious whether your team has conducted any experiments applying Gated Attention to Vision Transformers (ViT)? I implemented a Gated Multi-Head Attention module based on your paper and codebase:
> 
> import torch
> import torch.nn as nn
> import torch.nn.functional as F
> class GatedMultiHeadAttention(nn.Module):
>     def __init__(self, embed_dim, num_heads):
>         super().__init__()
>         self.embed_dim = embed_dim
>         self.num_heads = num_heads
>         self.head_dim = embed_dim // num_heads
>         
>         self.qkv = nn.Linear(embed_dim, 3 * embed_dim)
>         self.out_proj = nn.Linear(embed_dim, embed_dim)
>         
>         # Gating mechanism
>         self.gate = nn.Linear(embed_dim, embed_dim)
>         
>     def forward(self, x):
>         B, L, D = x.shape
>         
>         # 1. QKV projection
>         qkv = self.qkv(x).reshape(B, L, 3, self.num_heads, self.head_dim)
>         qkv = qkv.permute(2, 0, 3, 1, 4)
>         q, k, v = qkv[0], qkv[1], qkv[2]
>         
>         # 2. Scaled dot-product attention
>         attn_out = F.scaled_dot_product_attention(q, k, v)
>         # Shape: [B, num_heads, L, head_dim]
>         
>         # 3. Compute gate scores
>         gate_scores = torch.sigmoid(self.gate(x))
>         gate_scores = gate_scores.reshape(B, L, self.num_heads, self.head_dim)
>         gate_scores = gate_scores.permute(0, 2, 1, 3)
>         # Shape: [B, num_heads, L, head_dim]
>         
>         # 4. Apply gating
>         gated_out = attn_out * gate_scores
>         
>         # 5. Concatenate heads and project
>         gated_out = gated_out.permute(0, 2, 1, 3).reshape(B, L, D)
>         output = self.out_proj(gated_out)
>         
>         return output
> I tested this implementation on my own dataset, but didn't observe significant improvements. I would appreciate it if you could help verify whether my implementation is correct, or if there are any known considerations when applying this approach to vision tasks.
> 
> Thanks in advance!

Hello! Thank you for your question and the code you provided, it has inspired me, thank you!
I have read the paper, the official code, and your code "GatedMultiHeadAttention". I found an important small detail: how are the gating scores obtained? In Section 4.2(iii) of the paper, this sentence "**Specifically, SDPA output gating scores are derived from the hidden states corresponding to the current query (e.g. the Non-Linearity-Map in Eq 8 depends on Xi), whereas value gating scores are derived from hidden states associated with past keys and values (e.g. the Non-Linearity-Map in Eq 7 depends on each Xj).**" is very important. 

Specifically, at the code level, your code obtains the gating scores like this:

```
self.gate = nn.Linear(embed_dim, embed_dim)
gate_scores = torch.sigmoid(self.gate(x))
```
And the official code logic is as follows:

```
#Half is used to generate query vectors (num_heads * head_dim) and the other half is used to generate gating scores (num_heads * head_dim)

if self.elementwise_attn_output_gate:
    self.q_proj = nn.Linear(self.hidden_size, self.num_heads * self.head_dim * 2, bias=config.qkv_bias)

#Separate query and gating score
#The first half is the query vectorThe second half is the gating score

elif self.elementwise_attn_output_gate:
    query_states = query_states.view(bsz, q_len, self.num_key_value_heads, -1)
    query_states, gate_score = torch.split(query_states, 
                                           [self.head_dim * self.num_key_value_groups,
                                            self.head_dim * self.num_key_value_groups], dim=-1)
    gate_score = gate_score.reshape(bsz, q_len, -1, self.head_dim)
    query_states = query_states.reshape(bsz, q_len, -1, self.head_dim).transpose(1, 2)
```

The difference is that the official gating score is calculated based on the hidden state corresponding to the current query, so it is query-dependent.

The above are my personal views, welcome to discuss, I hope it is helpful. If I have misunderstood anything, please also point it out. Thanks again!

**Comment by qiuzh20** [comment]
Thanks for your question!

Both
```
    self.qkv = nn.Linear(embed_dim, 3 * embed_dim)
    self.out_proj = nn.Linear(embed_dim, embed_dim)
    
    # Gating mechanism
    self.gate = nn.Linear(embed_dim, embed_dim)
```
and

```
self.q_proj = nn.Linear(self.hidden_size, self.num_heads * self.head_dim * 2, bias=config.qkv_bias)
```

are conceptually correct. 

In our implementations, we decouple embed_dim with num_heads and head_dim, so embed_dim might not equals to um_heads * head_dim.
Both gating scores are calculated from the hidden states corresponding to the current query.

Moreover, in our training codebase, we use

```
self.qkv_proj = nn.Linear(self.hidden_size, self.num_heads * self.head_dim * 2 + self.num_key_value_heads * self.head_dim * 2, bias=config.qkv_bias)
```

and split the output into `q, gate_score, k, v` for better training efficiency.

**Comment by Zhangquan-se** [comment]
Thank you for your detailed response!
You are right!Indeed,gating scores are calculated from the hidden states corresponding to the current query.
Thank you for providing the training code repository, which increased my knowledge—I learned about MHA and GQA. You're using GQA, right? This line of code is very concise and efficient, without needing to write multiple Linear layers.Your response has given me new ideas for writing code.
```
self.qkv_proj = nn.Linear(self.hidden_size, self.num_heads * self.head_dim * 2 + self.num_key_value_heads * self.head_dim * 2, bias=config.qkv_bias)
```
Thank you again for your thoughtful response!
